### DISCUSSION TITLE
Shipment and Carton prefix

### DISCUSSION THREAD

**Discussion by tmtrademarked** [solidusio/solidus#3939] [Troubleshooting]
Hey, all - I'm trying to set up a new store, and I would like to use a different length/prefix for my orders/shipments/cartons `number` field. It looks like customizing order numbers is trivial and well supported using the Order::NumberGenerator class - but Shipments and Cartons are not currently set up to support that.

So this leads me to two questions:
1) What's the most idiomatic way to achieve this today? It's not honestly clear to me if monkey-patching these classes is a reasonable approach here, so I'm curious what folks think.
2) It seems like there's no good reason that Carton and Shipment shouldn't share similar logic to Orders with a configurable generator class. Is this something that has been discussed previously?

Thanks!

**Comment by jarednorman** [comment]
We've actually got three strategies in the codebase at this point:

1. Configurable number generator on `Spree::Order`
2. Permalink code used by `Spree::Carton` and `Spree::Shipment`
3. `generate_number` method on `Spree::ReturnAuthorization` and `Spree::CustomerReturn`

So, to answer your questions:

1. Set `Spree::Shipment.permalink_options[:prefix] = "YOURPREFIX"`. You may need to do this in a decorator rather than an initializer to get around code loading issues, not 100% on that.
2. I do believe people more frequently need to customize order numbers than any of the other things with numbers, but I think at _minimum_ all the other things should rely on `Spree::Core::Permalinks` unless there's something I'm missing.

**Reply by aldesantis** [reply]
It would definitely be interesting to see a unified API for "number generation" across all of Solidus, and it would probably not even be too complicated to do it in a backwards-compatible way by ensuring the new API mimics the old behavior.

**Reply by jarednorman** [reply]
Agreed. Fundamentally all the number generation is functionally the same. They could all use differently configured instances of the order number generator with only some small tweaks.
